src/annotator/frame-observer.js

```diff
@@ -65,22 +53,71 @@ export default class FrameObserver {
    */
   _removeFrame(frame) {
     this._onFrameRemoved(frame);
-
-    // Remove the frame from our list
-    this._handledFrames = this._handledFrames.filter(x => x !== frame);
+    this._handledFrames.delete(frame);
   }
 
   _discoverFrames() {
-    let frames = FrameUtil.findFrames(this._element);
+    let frames = findFrames(this._element);
 
     for (let frame of frames) {
-      if (!this._handledFrames.includes(frame)) {
+      if (!this._handledFrames.has(frame)) {
         this._addFrame(frame);
       }
     }
 
-    for (let frame of difference(this._handledFrames, frames)) {
-      this._removeFrame(frame);
+    for (let frame of this._handledFrames) {
+      if (!frames.has(frame)) {
+        this._removeFrame(frame);
+      }
     }
   }
 }
+
+/**
+ * Check if we can access this iframe's document
+ *
+ * @param {HTMLIFrameElement} iframe
+ */
+function isAccessible(iframe) {
+  try {
+    return !!iframe.contentDocument;
+  } catch (e) {
+    return false;
+  }
+}
+
+/**
+ * Resolves a Promise when the iframe's DOM is ready (loaded and parsed)
+ *
+ * @param {HTMLIFrameElement} iframe
+ * @return {Promise<void>}
+ */
+export function onDocumentReady(iframe) {
+  return new Promise(resolve => {
+    const iframeDocument = /** @type {Document} */ (iframe.contentDocument);
+    if (iframeDocument.readyState === 'loading') {
+      iframeDocument.addEventListener('DOMContentLoaded', () => {
+        resolve();
+      });
+    } else {
+      resolve();
```

Comment:
I wondered whether calling `resolve` in all cases where `readyState` is not `loading` was appropriate. The only other states are `interactive` or `complete` though, so that's OK.

Reply:
I discover an 'uninitialized' `readyState` in Firefox and other subtleties. See https://github.com/hypothesis/client/pull/3686